romancal/datamodels/library.py

```diff
@@ -0,0 +1,504 @@
+                filename = model.meta.filename
+                if filename in filenames:
+                    raise ValueError(
+                        f"Models in library cannot use the same filename: {filename}"
+                    )
+                self._model_store[index] = model
+                # FIXME: output models created during resample (during outlier detection
+                # an possibly others) do not have meta.observation which breaks the group_id
+                # code
+                try:
+                    group_id = _model_to_group_id(model)
+                except AttributeError:
+                    group_id = str(index)
+                # FIXME: assign the group id here as it may have been computed above
+                # this is necessary for some tweakreg tests that pass in a list of models that
+                # don't have group_ids. If this is something we want to support there may
+                # be a cleaner way to do this.
+                model.meta["group_id"] = group_id
+                members.append(
+                    {
+                        "expname": filename,
+                        "exptype": getattr(model.meta, "exptype", "SCIENCE"),
+                        "group_id": group_id,
+                    }
+                )
+
+            # make a fake association
+            self._asn = {
+                # TODO other asn data?
+                "products": [
+                    {
+                        "members": members,
+                    }
+                ],
+            }
+            self._members = self._asn["products"][0]["members"]
+
+            # _asn_dir?
+            # _asn_path?
+
+        elif isinstance(init, self.__class__):
+            # TODO clone/copy?
+            raise NotImplementedError()
+        else:
+            raise NotImplementedError()
+
+        # make sure first model is loaded in memory (as expected by stpipe)
+        if self._asn_n_members == 1:
+            # FIXME stpipe also reaches into _models (instead of _model_store)
+            self._models = [self._load_member(0)]
+
+    def __del__(self):
+        # FIXME when stpipe no longer uses '_models'
+        if hasattr(self, "_models"):
+            self._models[0].close()
+
+    @property
+    def asn(self):
+        # return a "read only" association
+        def _to_read_only(obj):
+            if isinstance(obj, dict):
+                return MappingProxyType(obj)
+            if isinstance(obj, list):
+                return tuple(obj)
+            return obj
+
+        return asdf.treeutil.walk_and_modify(self._asn, _to_read_only)
+
+    # TODO we may want to not expose this as it could go out-of-sync
+    # pretty easily with the actual models.
+    # @property
+    # def members(self):
+    #     return self.asn['products'][0]['members']
+
+    @property
+    def group_names(self):
+        names = set()
+        for member in self._members:
+            names.add(member["group_id"])
```

Comment:
Thanks for pointing this out. I missed that this is different than jwst (which allows setting group id in the association). I believe this was added in part to allow users to define grouping for tweakreg, skymatch, resample (any step that uses group id). It's more efficient to have this group id in the association instead of loading each model to compute the group id (which is why the group id is now being added to the [level 3 association generation for jwst](https://github.com/spacetelescope/jwst/issues/8523)).

The format here should match the format defined by the [current model container](https://github.com/spacetelescope/romancal/blob/ed6187ffa790fc6a37dd96f7365ba1ef8261fa74/romancal/datamodels/container.py#L499) and used by tweakreg, skymatch and resample in romancal. I can remove the part of this code that allows the group id to be defined from the association if roman isn't planning on supporting that feature. However the library does need to define the group id for the mentioned steps to work. Is there something incorrect in the format? If so would you provide an example?